### ISSUE TITLE
g++ produces broken executables with -O2 (pubsetbuf)

### ISSUE THREAD

**Issue by starius** [mxe/mxe#964]
Code:
```cpp
#include <fstream>

const int BUFFER_SIZE = 4096;
char buffer_[BUFFER_SIZE];

struct BufferedIfstream : public std::ifstream {
    BufferedIfstream() {
        rdbuf()->pubsetbuf(buffer_, BUFFER_SIZE);
    }
};

int main() {
    BufferedIfstream fff;
    return 0;
}
```

Compilation command:
```bash
$ ./usr/bin/i686-w64-mingw32.static-g++ mingw-pubsetbuf-bug.cpp -O2 -o mingw-pubsetbuf-bug.exe
```

Run:
```bash
$ wine mingw-pubsetbuf-bug.exe
Application tried to create a window, but no driver could be loaded.
Make sure that your X server is running and that $DISPLAY is set correctly.
wine: Unhandled page fault on write access to 0x7edd0000 at address 0x4951b8 (thread 0009), starting debugger...
Application tried to create a window, but no driver could be loaded.
Make sure that your X server is running and that $DISPLAY is set correctly.
Unhandled exception: page fault on write access to 0x7edd0000 in 32-bit code (0x004951b8).
Register dump:
 CS:0023 SS:002b DS:002b ES:002b FS:0063 GS:006b
 EIP:004951b8 ESP:00d0fc00 EBP:00d0fd80 EFLAGS:00010202(  R- --  I   - - - )
 EAX:ffffffff EBX:7edd0000 ECX:00d0fd5c EDX:00d0fd14
 ESI:00111e38 EDI:00000015
Stack dump:
0x00d0fc00:  00000000 00d0fbc0 0000000c 00000001
0x00d0fc10:  00111e38 00000001 00000001 004afca1
0x00d0fc20:  004bc844 0000ffff 00d0fc58 7bc3ef72
0x00d0fc30:  00000040 7bc36596 00d0fc60 00000000
0x00d0fc40:  7bc36686 0000000d 00000000 7bc36686
0x00d0fc50:  00d0fc70 004af400 004b01a0 00d0fd70
Backtrace:
=>0 0x004951b8 in mingw-bug (+0x951b8) (0x00d0fd80)
  1 0x004afca1 in mingw-bug (+0xafca0) (0x00d0fd80)
  2 0x7b83a4c6 SetUnhandledExceptionFilter+0x5() in kernel32 (0x00000000)
  3 0x004013e2 in mingw-bug (+0x13e1) (0x00000001)
  4 0x004013e2 in mingw-bug (+0x13e1) (0x00000001)
...
```
Full output: https://gist.github.com/starius/1832c17d46e8912c392c

Of course, the bug is reproducible on Windows as well. I used 32 bit Windows XP. 

Without `-O2` (or `-O3`) the bug disappears.

Without a class the bug disappears. The following code works correctly:
```cpp
#include <fstream>

const int BUFFER_SIZE = 4096;
char buffer_[BUFFER_SIZE];

int main() {
    std::ifstream fff;
    fff.rdbuf()->pubsetbuf(buffer_, BUFFER_SIZE);
    return 0;
}
```

Location of `char buffer_` doesn't matter: stack, heap, static memory.

The issue was introduced by commit da82d07dbeca23087aa4351914cf4b890c1d3972
Changing the version of `isl` from 0.14 back to 0.12.2 doesn't help. Re-enabling `cloog` doesn't help. Probably the bug is caused by  `gcc` version changing (4.9.2->5.2.0) or by some changes in `gcc-1.fixes`.

The issue seems to be related to the mailing thread about -O3: [stack smashing on i686 with -O3 since da82d07dbe](http://lists.nongnu.org/archive/html/mingw-cross-env-list/2015-10/msg00020.html).

**Comment by tonytheodore** [comment]
`-fno-devirtualize` seems to stop the crash.

**Comment by starius** [comment]
Confirmed.
  * `-O1 -fno-devirtualize` - no bug
  * `-O2 -fno-devirtualize` - no bug
  * `-O3 -fno-devirtualize` - no bug
  * `-O1 -fdevirtualize` - bug!
  * `-O1` - no bug
  * `-O1 -fno-devirtualize` - no bug!

There is the bug with devirtualization in gcc 4.9.0:
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=60965
It is fixed in gcc 5. I can't even find this code in gcc sources (5.2.0).

**Comment by starius** [comment]
Reproducible on Debian Stretch's GCC 5.2.1-22 (Linux arm64).

**Comment by starius** [comment]
The bug has been reported to gcc team: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=68175

**Comment by tonytheodore** [comment]
Nice one!

**Comment by starius** [comment]
[The bug](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=68175) has been marked as a duplicate of [another bug](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=67056).

I can reproduce the error with `-fsanitize=undefined`
```
runtime error: execution reached a __builtin_unreachable() call
```

By the way I can't compile in MXE with `-fsanitize=undefined`:
```
$ ./build.sh 
$ i686-w64-mingw32.static-g++ -fsanitize=undefined -O2 mingw-bug.cpp -o mingw-bug.exe
i686-w64-mingw32.static-ld: cannot find -lubsan
collect2: error: ld returned 1 exit status
```

That is why I can't check `-fsanitize=undefined` [the problem from the mailing list](http://lists.nongnu.org/archive/html/mingw-cross-env-list/2015-10/msg00020.html) (stack smashing on i686 with -O3 since da82d07dbe). 

Is it possible to enable this flag in MXE's gcc?

**Comment by starius** [comment]
I have [back-ported the patch](https://github.com/LuaAndC/mxe/commit/5c47e8ef678ea785b1092272fd3adb51beebff4e) and it doesn't work.

By the way I can't reproduce a crash with test file proposed on

**Comment by starius** [comment]
The patch [proposed](https://gcc.gnu.org/viewcvs/gcc/trunk/gcc/ipa-polymorphic-call.c?r1=229148&r2=229147&pathrev=229148) in gcc bug [67056](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=67056) doesn't fix the problem with class BufferedIfstream for Debian Stretch's gcc as well as for MXE. I have reopened gcc bug [68175](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=68175).

**Comment by starius** [comment]
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=68175
> It is fixed in Debian Stretch, gcc 5.3.1 20160101 (Debian 5.3.1-5).
> It is fixed in MinGW 5.3.0.

I've updated `plugins/gcc52/gcc52-overlay.mk` locally to use gcc 5.3.0. It compiled the test code to a valid binary. I'm compiling the rest of MXE with gcc 5.3.0. Maybe we should rename `plugins/gcc52/gcc52-overlay.mk` to `plugins/gcc5/gcc5-overlay.mk` and update gcc to 5.3.0 in it?

**Comment by tonytheodore** [comment]
> Maybe we should rename plugins/gcc52

Or:
plugins/gcc/gcc52
plugins/gcc/gcc53
plugins/gcc/gcc...
